src/handlers/payout/action.ts

```diff
@@ -68,18 +129,35 @@ export const handleIssueClosed = async () => {
     );
   }
 
+  // if bounty hunter has any penalty then deduct it from the bounty
+  const penaltyAmount = await getPenalty(assignee.login, payload.repository.full_name, paymentToken, chainId.toString());
+  if (penaltyAmount.gt(0)) {
+    logger.info(`Deducting penalty from bounty`);
+    const bountyAmount = ethers.utils.parseUnits(priceInEth, 18);
+    const bountyAmountAfterPenalty = bountyAmount.sub(penaltyAmount);
+    if (bountyAmountAfterPenalty.lte(0)) {
+      await removePenalty(assignee.login, payload.repository.full_name, paymentToken, chainId.toString(), bountyAmount);
+      const msg = `Permit generation skipped because bounty amount after penalty is 0`;
```

Comment:
I'd like to discuss this implementation more. Can you explain how far off we are from the ideal setup (below)?

Scenario: what exactly happens in the situation where a hunter's debt is $100, but a bounty amount is $50? 

Ideally we should be able to store their debt in a database, and they can work towards "paying the debt back" by completing new bounties, e.g. 2x $50 bounties. 

@rndquu roc

Reply:
In current implementation the debt is stored in the database so if the debt is 100$ and bounty is 50$, then 50$ will be removed from the debt. Then if they complete another 100$ bounty, 50$ will be removed again so debt will be 0$ and they will get 50$ permit.

Currently the debt is unique per repo and token/network, so if the token/network changes, it will not apply penalty to a new permit (we need off-chain oracle for that).

Reply:
> Currently the debt is unique per repo and token/network, so if the token/network changes, it will not apply penalty to a new permit

Isn't it better to remove `network_id` and `token_address` fields from the `penalty` DB table? This way when network or token is updated then penalty is applied anyway.

> Ideally we should be able to store their debt in a database, and they can work towards "paying the debt back" by completing new bounties, e.g. 2x $50 bounties.

So right now bounty hunter can get 0 reward. Perhaps it is better (as a part of another issue) to add some rule like "penalty can not exceed 50% of the reward"?

Reply:
> Isn't it better to remove `network_id` and `token_address` fields from the `penalty` DB table? This way when network or token is updated then penalty is applied anyway.

If network or token changes and it has a different token-to-USD exchange rate, then the penalty might be higher or lower in terms of USD. This could work if all tokens were pegged to USD but as @pavlovcik said that's not a guarantee